### ISSUE TITLE
false in definite integral of rational

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2661]
Imported from SourceForge on 2024-07-06 17:12:36
Created by **[willisbl](https://sourceforge.net/u/willisbl/)** on 2018-03-21 17:49:13
Original: https://sourceforge.net/p/maxima/bugs/3413

---

~~~~
(%i5)	integrate(1/(x^8+x+1),x,0,1);
(%o5)	(3^(3/2)*false-3^(3/2)*log(3)+4*%pi)/(19*3^(3/2))

(%i6)	build_info();
(%o6)	build_info(version="5.40.0",timestamp="2017-06-01 18:54:43",host="x86_64-w64-mingw32",lisp_name="SBCL",lisp_version="1.3.15")
~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:37
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2018-04-06 19:48:49
Original: https://sourceforge.net/p/maxima/bugs/3413/#9381

---

Looks like the problem is in ZMTORAT.

~~~~
(%i1) trace(defint, ?zmtorat) $            

(%i2) integrate(1/(x^8+x+1),x,0,1);
                    1
1 Enter defint [----------, x, 0, 1]
                 8
                x  + x + 1
                    6       5        4        3        2
 1 Enter zmtorat [yx  + 6 yx  + 15 yx  + 20 yx  + 15 yx  + 6 yx + 1, 
    8        7        6        5         4        3        2
3 yx  + 15 yx  + 49 yx  + 91 yx  + 105 yx  + 77 yx  + 35 yx  + 9 yx + 1, 8, 
#<FUNCTION ZTORAT>]
                  3/2          3/2
                 3    false - 3    log(3) + 4 %pi
 1 Exit  zmtorat --------------------------------
                                 3/2
                             19 3
                3/2          3/2
               3    false - 3    log(3) + 4 %pi
1 Exit  defint --------------------------------
                               3/2
                           19 3
                        3/2          3/2
                       3    false - 3    log(3) + 4 %pi
(%o2)                  --------------------------------
                                       3/2
                                   19 3
~~~~

My current theory is that ZMTORAT is getting a NIL from a function it calls (ZTORAT) and not noticing that, and rolling the NIL into its result. Tracing `xthru` shows it gets the NIL as part of the intermediate result.

~~~~
(%i3) trace (?ztorat, xthru) $

(%i4) integrate(1/(x^8+x+1),x,0,1);
                    1
1 Enter defint [----------, x, 0, 1]
                 8
                x  + x + 1
                    6       5        4        3        2
 1 Enter zmtorat [yx  + 6 yx  + 15 yx  + 20 yx  + 15 yx  + 6 yx + 1, 
    8        7        6        5         4        3        2
3 yx  + 15 yx  + 49 yx  + 91 yx  + 105 yx  + 77 yx  + 35 yx  + 9 yx + 1, 8, 
#<CLOSURE (LAMBDA (&REST TRACE-ARGS) :IN MAKE-TRACE-HOOK) {C8CC685}>]
                                    2
  1 Enter ztorat [(- 6 yx) + 1, 3 yx  + 3 yx + 1, 2]
                 4 %pi
  1 Exit  ztorat ----- - log(3)
                  3/2
                 3
                      5        4        3        2
  1 Enter ztorat [2 yx  + 12 yx  + 48 yx  + 73 yx  + 60 yx + 18, 
                            6       5        4        3        2
                          yx  + 4 yx  + 12 yx  + 17 yx  + 14 yx  + 6 yx + 1, 6]
  1 Exit  ztorat false
                         4 %pi
                         ----- - log(3)
                          3/2
                 false   3
  1 Enter xthru [----- + --------------]
                  19           19
                  4 %pi
                  ----- - log(3)
                   3/2
                  3
   2 Enter xthru [--------------]
                        19
                   1
    3 Enter xthru [--]
                   19
                  1
    3 Exit  xthru --
                  19
                   4 %pi
    3 Enter xthru [----- - log(3)]
                    3/2
                   3
                    4 %pi
     4 Enter xthru [-----]
                     3/2
                    3
      5 Enter xthru [4]
      5 Exit  xthru 4
                      1
      5 Enter xthru [----]
                      3/2
                     3
       6 Enter xthru [3]
       6 Exit  xthru 3
                     1
      5 Exit  xthru ----
                     3/2
                    3
      5 Enter xthru [%pi]
      5 Exit  xthru %pi
                   4 %pi
     4 Exit  xthru -----
                    3/2
                   3
     4 Enter xthru [- log(3)]
      5 Enter xthru [- 1]
      5 Exit  xthru - 1
      5 Enter xthru [log(3)]
      5 Exit  xthru log(3)
     4 Exit  xthru - log(3)
                           3/2
                  4 %pi - 3    log(3)
    3 Exit  xthru -------------------
                          3/2
                         3
                          3/2
                 4 %pi - 3    log(3)
   2 Exit  xthru -------------------
                           3/2
                       19 3
                  false
   2 Enter xthru [-----]
                   19
                   1
    3 Enter xthru [--]
                   19
                  1
    3 Exit  xthru --
                  19
    3 Enter xthru [false]
    3 Exit  xthru false
                 false
   2 Exit  xthru -----
                  19
                 3/2          3/2
                3    false - 3    log(3) + 4 %pi
  1 Exit  xthru --------------------------------
                                3/2
                            19 3
                  3/2          3/2
                 3    false - 3    log(3) + 4 %pi
 1 Exit  zmtorat --------------------------------
                                 3/2
                             19 3
                3/2          3/2
               3    false - 3    log(3) + 4 %pi
1 Exit  defint --------------------------------
                               3/2
                           19 3
                        3/2          3/2
                       3    false - 3    log(3) + 4 %pi
(%o4)                  --------------------------------
                                       3/2
                                   19 3
~~~~

Not sure what's the correct result here. Maybe ZMTORAT should give up if ZTORAT returns NIL? Or maybe there is a problem in ZTORAT? I'm not sure what ZMTORAT and ZTORAT are supposed to do.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:41
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2018-04-06 20:56:08
Original: https://sourceforge.net/p/maxima/bugs/3413/#eada

---

I see ZTORAT concludes with:

~~~~
    (t
     ;; Let's not signal an error here.  Return nil so that we
     ;; eventually return a noun form if no other algorithm gives
     ;; a result.
     #+(or)
     (merror (intl:gettext "defint: keyhole integration failed.~%"))
     nil)))
~~~~

Commit bc727e85 says, in part:

~~~~
    o Don't signal an error in ZTORAT if keyhole integration fails.
      Instead, just return NIL, which should cause the integration routine
      to return a noun form eventually.
~~~~

So I think the error can be avoided by getting ZMTORAT to recognize the NIL result and give up in that case.

I wonder if there are other cases in which ZTORAT is called. ZMTORAT calls ZTORAT because ZTORAT is supplied to ZMTORAT as its argument FN1, so we have to be careful to look for calls like that, not just obvious ones like `(ztorat ...)`.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:44
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2018-04-26 18:54:25
Original: https://sourceforge.net/p/maxima/bugs/3413/#ea36

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 17:12:48
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2018-04-26 18:54:26
Original: https://sourceforge.net/p/maxima/bugs/3413/#e455

---

Fixed by commit 811e66c; `integrate(1/(x^8+x+1),x,0,1)` now returns a noun expression.

Commit 811e66c also fixes bug report https://sourceforge.net/p/maxima/bugs/2644/.
